README.md

```diff
@@ -1,138 +1,615 @@
+    return wrapped;
+  })
+}
+```
+
+In the above example, the *composed decorator* `@logged` is defined to expand out into a call of the `@wrap` decorator with a particular fixed callback.
+
+### `@defineElement`
+
+[HTML Custom Elements](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements) lets you define your own HTML element. Elements are registered using `customElements.define`. Using decorators, the registration can happen up-front:
+
+```mjs
+import { @defineElement } from "./defineElement.mjs";
+
+@defineElement('my-class')
+class MyClass extends HTMLElement { }
+```
+
+The `@defineElement` decorator is based on the `@register` decorator. This decorator is called when the class is finished being defined, and it calls the provided callback with the class that resulted.
+
+```mjs
+// defineElement.mjs
+export decorator @defineElement(name, options) {
+  @register(klass => customElements.define(name, klass, options))
+}
+```
+
+This example uses a *decorator arrow function* `@(args) => @decorators` which lets a decorator definition take arguments that can be used to supply arguments to other decorators in its definition.
+
+### `@metadata`
+
+The `@metdata(key, value)` decorator is similar to [`@Reflect.metadata`](https://github.com/rbuckton/reflect-metadata): It allows the easy retrival of information which was stored by annotating the class.
+
+```mjs
+import { @metadata } from "./metadata.mjs";
+
+// partially apply the decorator locally for terseness
+decorator @localMeta { @metadata("key", "value") }
+
+@localMeta class C {
+  @localMeta method() { }
+}
+
+Reflect.getMetadata(C, "key");                      // "value"
+Reflect.getMetadata(C.prototype, "key", "method");  // "value"
+```
+
+`@metadata` can also be defined in terms of `@register`. When `@register` is used with a public field, method or accessor, it is called with the second argument being the property key of that class eleemnt.
+
+```mjs
+// metadata.mjs
+import "reflect-metadata";
+
+export decorator @metadata(key, value) {
+  @register((target, prop) => Reflect.defineMetadata(key, value, target, prop))
+}
+```
+
+### `@frozen`
+
+The `@frozen` decorator freezes the constructor and its `prototype` so that they cannot be mutated after the class is defined. It leaves instances mutable. Example usage:
+
+```mjs
+import { @frozen } from "./frozen.mjs";
+
+@frozen
+class MyClass {
+  method() { }
+}
+
+MyClass.method = () => {};            // TypeError to add a method
+MyClass.prototype.method = () => {};  // TypeError to overwrite a method
+MyClass.prototype.method.foo = 1;     // TypeError to mutate a method
+```
+
+`@frozen` is implemented with the `@register` decorator, which allows a callback to be scheduled after the class is created. The callback is passed the class as an argument.
+
+```mjs
+// frozen.mjs
+export decorator @frozen {
+  @register(klass => {
+    Object.freeze(klass);
+    for (const [key, value] of Object.entries(klass)) {
```

Comment:
you'd probably want to use `Reflect.ownKeys` here, so it's not limited to enumerables and string keys.

Reply:
Fixed, thanks.